contrib/native/client/src/clientlib/saslAuthenticatorImpl.cpp

```diff
@@ -0,0 +1,211 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <vector>
+#include <boost/algorithm/string.hpp>
+#include <boost/assign.hpp>
+#include "saslAuthenticatorImpl.hpp"
+
+#include "drillClientImpl.hpp"
+#include "logger.hpp"
+
+namespace Drill {
+
+#define DEFAULT_SERVICE_NAME "drill"
+
+#define KERBEROS_SIMPLE_NAME "kerberos"
+#define KERBEROS_SASL_NAME "gssapi"
+#define PLAIN_NAME "plain"
+
+const std::map<std::string, std::string> SaslAuthenticatorImpl::MECHANISM_MAPPING = boost::assign::map_list_of
+    (KERBEROS_SIMPLE_NAME, KERBEROS_SASL_NAME)
+    (PLAIN_NAME, PLAIN_NAME)
+;
+
+boost::mutex SaslAuthenticatorImpl::s_mutex;
+bool SaslAuthenticatorImpl::s_initialized = false;
+
+SaslAuthenticatorImpl::SaslAuthenticatorImpl(const DrillUserProperties* const properties) :
+    m_properties(properties), m_pConnection(NULL), m_pwd_secret(NULL) {
+
+    if (!s_initialized) {
+        boost::lock_guard<boost::mutex> lock(SaslAuthenticatorImpl::s_mutex);
+        if (!s_initialized) {
+            // set plugin path if provided
+            if (DrillClientConfig::getSaslPluginPath()) {
+                char *saslPluginPath = const_cast<char *>(DrillClientConfig::getSaslPluginPath());
+                sasl_set_path(0, saslPluginPath);
+            }
+
+            // loads all the available mechanism and factories in the sasl_lib referenced by the path
+            const int err = sasl_client_init(NULL);
+            if (0 != err) {
+                std::stringstream errMsg;
+                errMsg << "Failed to load authentication libraries. code: " << err;
+                DRILL_MT_LOG(DRILL_LOG(LOG_TRACE) << errMsg << std::endl;)
```

Comment:
This would not compile. Try errMsg.str()
On windows, Im getting:
Error 443	error C2678: binary '<<' : no operator found which takes a left-hand operand of type 'std::ostream' (or there is no acceptable conversion)